src/shims/unix/linux/sync.rs

```diff
@@ -203,6 +205,10 @@ pub fn futex<'tcx>(
                 this.write_scalar(Scalar::from_machine_isize(-1, this), dest)?;
                 return Ok(());
             }
+            // Together with the SeqCst fence in futex_wait, this makes sure that futex_wait
+            // will see the latest value on addr which could be changed by our caller
+            // before doing the syscall.
+            this.atomic_fence(&[], AtomicFenceOp::SeqCst)?;
```

Comment:
So the issue is that previously, futex_wake didn't actually do anything on `addr`, so it was not ordered wrt other operations on `addr` -- and this extra fence here inserts it into the SeqCst order?

Reply:
futex wake syscall is not supposed to do anything on `addr`. The API user is supposed to change its value before invoking the wake syscall.

> When releasing the lock, a thread has to first reset the lock state to not acquired and then execute a futex operation

https://man7.org/linux/man-pages/man2/futex.2.html

Reply:
Okay so what is the correctness argument here for a racing wait and wake? We have two SC fences, so the thing weare allowed to do is case distinction on which fence comes first.

- In case the wake fence goes first, the wait fence will establish a happens-before to that fence, so the read in wait will definitely pick up all writes before the wake. Therefore it will not sleep and we are good.
- In case the wait fence goes first... then what? Since all of wait is atomic and cannot be preempted, we know the waiting thread will be in the futex wait set, so the waking thread will see and wake it. But this relies on all of wait (fence, read, add to wait set) being non-preemptible.

Reply:
Yes your reasoning are right for both cases. Except for that the fences don't actually establish happens-before, it's just a special property unique to SC fence (cannot be replicated with acquire/release fences):
```
T1:                                                 T2:
na = 10
X.store(42, Ordering.any)             
fence(sc)
                                                     fence(sc)
                                                     X.load(Ordering.any)
                                                     r = na // still a data race, does not happen before na = 10
``` 
The load _will_ read 42 as long as T2's fence and load are interleaved after T1's fence, but it does not establish HB with anything, though this doesn't matter for our case.

> For atomic operations A and B on an atomic object M, where A modifies M and B takes its value, if there is a memory_order_seq_cst fence X such that A is sequenced before X and B follows X in S, then B observes either the effects of A or a later modification of M in its modification order.

Reply:
Oh wow, that is wild. So it's a "definitely visible but not happens-before" edge? I have no idea how that makes any sense but it's good enough for our purposes here.

Can you please add a comment with this correctness argument to the code?

Reply:
Also, just to be sure I understand -- earlier you said "It doesn't matter if there is a preemption between the fence and read", but in my correctness argument we crucially use that fence+read are atomic. So it does matter, after all?

Reply:
You are right, the fence+read also have to be atomic. I've added the correctness argument as comments